gu-provider/src/server.rs

```diff
@@ -1,195 +1,335 @@
-use gu_persist::error::Error as ConfigError;
 use hdman::HdMan;
-use mdns::Responder;
-use mdns::Service;
-use std::borrow::Cow;
-use std::net::{SocketAddr, ToSocketAddrs};
-use std::sync::Arc;
+use std::{
+    net::{SocketAddr, ToSocketAddrs},
+    sync::Arc,
+};
 
 #[derive(Serialize, Deserialize, Clone)]
 #[serde(rename_all = "camelCase")]
-struct ServerConfig {
+pub(crate) struct ProviderConfig {
+    #[serde(default = "ProviderConfig::default_p2p_port")]
     p2p_port: u16,
     #[serde(skip_serializing_if = "Option::is_none")]
     control_socket: Option<String>,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    hub_addr: Option<SocketAddr>,
     #[serde(default)]
-    pub(crate) publish_service: bool,
+    pub(crate) hub_addrs: Vec<SocketAddr>,
+    #[serde(default)]
+    publish_service: bool,
+    #[serde(default = "ProviderConfig::default_connect_mode")]
+    pub(crate) connect_mode: ConnectMode,
 }
 
-impl Default for ServerConfig {
+impl Default for ProviderConfig {
     fn default() -> Self {
-        ServerConfig {
-            p2p_port: 61621,
+        ProviderConfig {
+            p2p_port: Self::default_p2p_port(),
             control_socket: None,
-            hub_addr: None,
+            hub_addrs: Vec::new(),
             publish_service: true,
+            connect_mode: Self::default_connect_mode(),
         }
     }
 }
 
-impl ServerConfig {
+impl ServerConfig for ProviderConfig {
+    fn port(&self) -> u16 {
+        Self::default_p2p_port()
+    }
+}
+
+pub(crate) type ProviderClient = ServerClient<ProviderConfig>;
+
+#[derive(Serialize, Deserialize, PartialEq, Clone, Message, Debug, Copy)]
+#[rtype(result = "Result<Option<()>, String>")]
+pub(crate) enum ConnectMode {
+    Auto,
+    Config,
+}
+
+impl ProviderConfig {
     fn p2p_addr(&self) -> impl ToSocketAddrs {
         ("0.0.0.0", self.p2p_port)
     }
+
+    fn default_p2p_port() -> u16 {
+        61621
+    }
+
+    fn default_connect_mode() -> ConnectMode {
+        ConnectMode::Config
+    }
 }
 
-impl HasSectionId for ServerConfig {
+impl HasSectionId for ProviderConfig {
     const SECTION_ID: &'static str = "provider-server-cfg";
 }
 
-pub struct ServerModule {
-    config_path: Option<String>,
-    hub_addr: Option<SocketAddr>,
-}
+pub struct ServerModule;
 
 impl ServerModule {
     pub fn new() -> Self {
-        ServerModule {
-            config_path: None,
-            hub_addr: None,
-        }
+        ServerModule
     }
 }
 
-impl Module for ServerModule {
-    fn args_declare<'a, 'b>(&self, app: clap::App<'a, 'b>) -> clap::App<'a, 'b> {
-        app.arg(
-                Arg::with_name("hub_addr")
-                   .short("a")
-                   .long("hub address")
-                   .takes_value(true)
-                   .value_name("IP:PORT")
-                   .help("IP and PORT of Hub to connect to")
-        )
-    }
+fn get_node_id(keys: Box<SafeEthKey>) -> NodeId {
+    let node_id = NodeId::from(keys.address().as_ref());
```

Comment:
while this is useful, it is not consistent with `gu-hub` behavior.
I'd remove from here it since `gu-ethkey` prints address already